File: umapwrapper.R


Rumap <- function(data, verbose = FALSE, tmpinput = NULL, tmpoutput = NULL, statsoutput = NULL,
                  n.neighbors=15, n.components =2, metric='euclidean',
                  gamma = 1.0, alpha = 1.0, min.dist = 0.1, spread = 1.0, debug = FALSE) {

    if (is.null(tmpinput))
        tmpinput <- tempfile();

    if (is.null(tmpoutput))
        tmpoutput <- tempfile();

    if (is.null(statsoutput))
        statsoutput <- tempfile();

    if (debug) {
        cat(paste0('tmpinput=',tmpinput),'\n');
        cat(paste0('tmpoutput=',tmpoutput), '\n');
        cat(paste0('statsoutput=',statsoutput),'\n');
    }
    
    if (verbose) cat('Writing data... ');
    write.table(data, file=tmpinput, quote = F, row.names = F, col.names =T)
    if (verbose) cat('done\n');

    if (verbose) cat('Running external umap script...');
    cmd.params <- paste0(
        ' --inputcsv ', tmpinput,
        ' --output ', tmpoutput,
        ' --statsfile ', statsoutput,
        ' --n_neighbors ', n.neighbors,
        ' --n_components ', n.components,
        ' --metric ', metric,
        ' --gamma ', gamma,
        ' --alpha ', alpha,
        ' --min_dist ', min.dist,
        ' --spread ', spread,
        ' '
    )

    exec.path <- './umapwrapper.py'
    
    cmd <- paste0(exec.path, cmd.params)

    if (debug) {
        cat('\n');
        cat(paste0('Executing command: \n', cmd, '\n'))
    }
    
    retval <- system(cmd)
    if (retval != 0) {
        stop(paste0('An error occured! Return value: ', retval));
    }
    
    if (verbose) cat('done\n')

    if (verbose) cat('Reading embedding...');
    emb <- read.csv(tmpoutput, header=F)
    rownames(emb) <- rownames(data)
    if (verbose) cat('done\n');

    if (verbose) cat('Reading stats...');
    runtime <- scan(statsoutput);
    if (verbose) cat('done\n')

    # Return value
    list(embedding=emb,runtime=runtime[1])
}

File: umapwrapper.py
#!/usr/bin/env python
import sys
import numpy
import umap
import argparse

# Parse command line
parser = argparse.ArgumentParser(description = 'Command line wrapper for umap')
parser.add_argument('-v', '--verbose', help='set verbosity', action='store_true')
parser.add_argument('--inputcsv', help='input file in csv format, rows are data points')
parser.add_argument('--output', help='output file that will contain the embedding')
parser.add_argument('--statsfile', default = None,  help='file to save run time statistics')

parser.add_argument('--n_neighbors', type=int, default=15, help='The size of the local neighborhood for manifold approximation')
parser.add_argument('--n_components', type=int, default=2, help='The dimension of the sapce to embed into.')
parser.add_argument('--metric', default='euclidean', help='The metric to use to compute distances in high dimensional space')
parser.add_argument('--gamma', default=1.0, type=float,help='Weighting applied to negative samples in low dimensional embedding optimization.')
parser.add_argument('--n_edge_samples', type=int, default=None, help='The number of edge/1-simplex samples to be used in optimizing the low dimensional embedding')
parser.add_argument('--alpha', default=1.0, type=float,help='Initial learning rate')
parser.add_argument('--min_dist', default=0.1, type=float,help='The effective minimum distance between embedded points')
parser.add_argument('--spread', default=1.0, type=float,help='The effective scale of embedded points')

# Currently not implemented in this script
#parser.add_argument('--init', default='spectral')
#parser.add_argument('--a', default=None)
#parser.add_argument('--b', default=None)
#parser.add_argument('--metric_kwds')

args = parser.parse_args()

# Debug
print(args)
#sys.exit();

inputfile = args.inputcsv
outputfile = args.output

# Load the input data, ignoreing header line
inputdata = numpy.loadtxt(inputfile, skiprows=1)

#Run umap
import time

t0 = time.time()


emb = umap.UMAP(alpha = args.alpha, gamma = args.gamma, metric = args.metric, min_dist = args.min_dist, n_components = args.n_components, n_neighbors = args.n_neighbors, spread = args.spread).fit_transform(inputdata)

t1 = time.time()

umap_runtime = t1 - t0

if args.statsfile is not None:
    with open(args.statsfile, 'w') as f:
        f.write('%d' % umap_runtime)

# Save the outputfile
numpy.savetxt(outputfile, emb, delimiter = ",")
